OpenRA.Mods.D2k/Traits/Sandworm.cs

```diff
@@ -0,0 +1,145 @@
+
+		[Desc("The range at which the worm launches an attack regardless of noise levels.")]
+		public readonly WRange IgnoreNoiseAttackRange = WRange.FromCells(3);
+
+		[Desc("The chance this actor has of disappearing after it attacks (in %).")]
+		public readonly int ChanceToDisappear = 80;
+
+		[Desc("Name of the sequence that is used when the actor is idle or moving (not attacking).")]
+		public readonly string IdleSequence = "idle";
+
+		public override object Create(ActorInitializer init) { return new Sandworm(init.Self, this); }
+	}
+
+	class Sandworm : Wanders, ITick, INotifyKilled
+	{
+		public readonly SandwormInfo Info;
+
+		readonly Lazy<Mobile> mobile;
+		readonly Lazy<WormManager> manager; 
+		readonly Lazy<RenderUnit> renderUnit;
+		readonly Lazy<AttackBase> attackTrait;
+
+		public bool IsMovingTowardTarget { get; private set; }
+
+		public bool IsAttacking;
+
+		int targetCountdown;
+
+		public Sandworm(Actor self, SandwormInfo info)
+			: base(self, info)
+		{
+			Info = info;
+			mobile = Exts.Lazy(self.Trait<Mobile>);
+			renderUnit = Exts.Lazy(self.Trait<RenderUnit>);
+			attackTrait = Exts.Lazy(self.Trait<AttackBase>);
+			manager = Exts.Lazy(() => self.World.WorldActor.Trait<WormManager>());
+		}
+
+		public override void OnBecomingIdle(Actor self)
+		{
+			if (renderUnit.Value.DefaultAnimation.CurrentSequence.Name != Info.IdleSequence)
+				renderUnit.Value.DefaultAnimation.ReplaceAnim("idle");
+
+			base.OnBecomingIdle(self);
+		}
+
+		public override void DoAction(Actor self, CPos targetCell)
+		{
+			IsMovingTowardTarget = false;
+
+			RescanForTargets(self);
+
+			if (IsMovingTowardTarget)
+				return;
+
+			self.QueueActivity(mobile.Value.MoveWithinRange(Target.FromCell(self.World, targetCell, SubCell.Any), WRange.FromCells(1)));
+		}
+
+		public void Tick(Actor self)
+		{
+			if (--targetCountdown > 0 || IsAttacking || !self.IsInWorld)
+				return;
+
+			RescanForTargets(self);
+		}
+
+		void RescanForTargets(Actor self)
+		{
+			targetCountdown = Info.TargetRescanInterval;
+
+			var actorsInRange = self.World.FindActorsInCircle(self.CenterPosition, Info.MaxSearchRadius);
+			var noiseDirection = WVec.Zero;
+
+			foreach (var actor in actorsInRange)
+			{
+				if (!actor.IsInWorld)
+					continue;
+
+				var noise = actor.TraitOrDefault<AttractsWorms>();
+				if (noise == null)
+					continue;
+
+				// TODO: Test if we really want to ignore actors that are on rock
+				if (!mobile.Value.CanEnterCell(actor.Location, null, false))
+					continue;
+
+				var distance = actor.CenterPosition - self.CenterPosition;
+
+				// If close enough, we don't care about other actors
+				if (distance.Length <= Info.IgnoreNoiseAttackRange.Range)
+				{
+					self.CancelActivity();
+					attackTrait.Value.ResolveOrder(self, new Order("Attack", actor, true) { TargetActor = actor });
+					return;
+				}
+
+				noiseDirection += noise.Info.Intensity * distance / noise.Info.Attenuation;
```

Comment:
You'll need to divide through by `distance.LengthSquared` to give a 1/r falloff, or `distance.Length * 3` for a more physically correct 1/r^2